### PR TITLE
Unikernel and single executable feeder & witness

### PR DESCRIPTION
**Pull request by mhutchinson**
This runs the omniwitness feeder & witness, but purely in-memory.
Until this is power cycled, this performs all of the same work that
the docker omniwitness does (except for pushing the checkpoints to github).

`make CROSS_COMPILE=arm-none-eabi- imx` builds the imx file

Note that this disables TLS and performs manual DNS in order to support
network operations on a unikernel running on the usbarmory.

Note that this uses build tags to keep the rest of the tree free from
tamago-specific build concerns. This means `go build ./...` still works
 at the root, for example.

### CODE REVIEW

**Comment by mhutchinson** [comment]
Hi @AlCutter, this is ready for review!

File: witness/golang/omniwitness/internal/omniwitness/omniwitness.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	"github.com/google/trillian-examples/internal/feeder"
+	"github.com/google/trillian-examples/internal/feeder/pixelbt"
+	"github.com/google/trillian-examples/internal/feeder/rekor"
+	"github.com/google/trillian-examples/internal/feeder/serverless"
+	"github.com/google/trillian-examples/internal/feeder/sumdb"
+)
+
+const (
+	// Interval between attempts to feed checkpoints
+	feedInterval = 5 * time.Minute
```
Comment (AlCutter) [inline witness/golang/omniwitness/internal/omniwitness/omniwitness.go:51]: Worth a TODO to consider making this configurable?

File: witness/golang/omniwitness/internal/omniwitness/omniwitness.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+// singleLogFeederConfig encapsulates the feeder config for a feeder that can only
+// feed a single log.
+type singleLogFeederConfig struct {
+	// Log defines the source log to feed from.
+	Log config.Log `yaml:"Log"`
+}
+
+// multiLogFeederConfig encapsulates the feeder config for a feeder that can only
+// feed a single log.
```
Comment (AlCutter) [inline witness/golang/omniwitness/internal/omniwitness/omniwitness.go:62]: Should this comment be updated?

File: witness/golang/omniwitness/internal/omniwitness/omniwitness.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+	// This error group will be used to run all top level processes
+	g := errgroup.Group{}
+
+	type logFeeder func(context.Context, config.Log, feeder.Witness, *http.Client, time.Duration) error
+	feeders := make(map[config.Log]logFeeder)
+	// Feeder: SumDB
+	sumdbConfig := config.Log{
+		Origin:    "go.sum database tree",
+		PublicKey: "sum.golang.org+033de0ae+Ac4zctda0e5eza+HJyk9SxEdh+s3Ux18htTTAD8OuAn8",
+	}
```
Comment (AlCutter) [inline witness/golang/omniwitness/internal/omniwitness/omniwitness.go:82]: Feels like this should be in the config file too, so it's not a bit different/special?

File: witness/golang/omniwitness/internal/omniwitness/omniwitness.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+	// Calculate the ID (this doesn't happen if the struct is not parsed from yaml)
+	sumdbConfig.ID = log.ID(sumdbConfig.Origin, []byte(sumdbConfig.PublicKey))
+	feeders[sumdbConfig] = sumdb.FeedLog
+
+	// Feeder: PixelBT
+	pixelFeederConfig := singleLogFeederConfig{}
+	if err := yaml.Unmarshal(omniwitness.ConfigFeederPixel, &pixelFeederConfig); err != nil {
+		return fmt.Errorf("failed to unmarshal config: %v", err)
```
Comment (AlCutter) [inline witness/golang/omniwitness/internal/omniwitness/omniwitness.go:90]: nit - might be worth updating these errors to include the config file/var name?

File: witness/golang/omniwitness/monolith/monolith.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	"time"
+
+	"github.com/golang/glog"
+	i_omni "github.com/google/trillian-examples/witness/golang/omniwitness/internal/omniwitness"
+	"golang.org/x/mod/sumdb/note"
+)
+
+const (
+	// Timeout for any http requests.
+	httpTimeout = 10 * time.Second
```
Comment (AlCutter) [inline witness/golang/omniwitness/monolith/monolith.go:33]: nit: worth flagifying this?

File: witness/golang/omniwitness/monolith/monolith.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	"context"
+	"flag"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/golang/glog"
+	i_omni "github.com/google/trillian-examples/witness/golang/omniwitness/internal/omniwitness"
```
Comment (AlCutter) [inline witness/golang/omniwitness/monolith/monolith.go:27]: Do you need the rename? (maybe this was a hangover from before-times?)

File: witness/golang/omniwitness/monolith/monolith.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	signingKey = flag.String("private_key", "", "The note-compatible signing key to use")
+)
+
+func main() {
+	flag.Parse()
+	ctx := context.Background()
+
+	httpListener, err := net.Listen("tcp", *addr)
+	if err != nil {
+		glog.Fatalf("failed to listen on 8080")
```
Comment (AlCutter) [inline witness/golang/omniwitness/monolith/monolith.go:47]: nit: `"...listen on %q", *addr)`

File: witness/golang/omniwitness/usbarmory/Makefile [hunk 8]
```diff
@@ -0,0 +1,97 @@
+# http://github.com/f-secure-foundry/tamago-example
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/Makefile:1]: super nit: They just moved from f-secure-foundry -> usbarmory git hub orgs, could update here?

File: witness/golang/omniwitness/usbarmory/Makefile [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	$(QEMU) -kernel $(APP) -S -s
+
+#### dependencies ####
+
+$(APP): check_tamago
+	$(GOENV) $(TAMAGO) build $(GOFLAGS) -o ${APP}
+
+$(APP).dcd: check_tamago
+$(APP).dcd: GOMODCACHE=$(shell ${TAMAGO} env GOMODCACHE)
+$(APP).dcd: TAMAGO_PKG=$(shell grep "github.com/f-secure-foundry/tamago v" ../../../../go.mod | awk '{print $$1"@"$$2}')
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/Makefile:82]: Does this gh org need changing too?

File: witness/golang/omniwitness/usbarmory/main.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+	"context"
+	"flag"
+	"time"
+
+	"github.com/golang/glog"
+	i_omni "github.com/google/trillian-examples/witness/golang/omniwitness/internal/omniwitness"
+	"golang.org/x/mod/sumdb/note"
+	"golang.org/x/sync/errgroup"
+
+	_ "github.com/f-secure-foundry/tamago/board/f-secure/usbarmory/mark-two"
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/main.go:34]: `github.com/usbarmory/tamago/...`

File: witness/golang/omniwitness/usbarmory/main.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+// the imx file to the device.
+package main
+
+import (
+	"context"
+	"flag"
+	"time"
+
+	"github.com/golang/glog"
+	i_omni "github.com/google/trillian-examples/witness/golang/omniwitness/internal/omniwitness"
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/main.go:30]: Do you need this rename?

File: witness/golang/omniwitness/usbarmory/main.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+	// Generated from https://go.dev/play/p/uWUKLNK6h9v
+	// TODO(mhutchinson): these need to be read from file instead of constants
+	publicKey  = "TrustMe+68958214+AQ4Ys/PsXqfhPkNK7Y7RyYUMOJvfl65PzJOEiq9VFPjF"
+	signingKey = "PRIVATE+KEY+TrustMe+68958214+AZKby3TDZizdARF975ZyLJwGbHTivd+EqbfYTN5qr2cI"
+	_          = publicKey // public key is here so it doesn't get lost, we don't need it right now.
+)
+
+func main() {
+	flag.Parse()
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/main.go:49]: Do you need to import `flag` and run `Parse` here?

File: witness/golang/omniwitness/usbarmory/main.go [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+	ctx := context.Background()
+	// This error group will be used to run all top level processes
+	g := errgroup.Group{}
+
+	httpListener, err := initNetworking()
+	if err != nil {
+		glog.Exitf("failed to init usb networking: %v", err)
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/main.go:56]: super nit: capital "F" to match `Exit`s below

File: witness/golang/omniwitness/usbarmory/usbnet.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
+// https://github.com/f-secure-foundry/tamago-example
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/usbnet.go:1]: GH org

File: witness/golang/omniwitness/usbarmory/usbnet.go [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	usbnet "github.com/f-secure-foundry/imx-usbnet"
+	"github.com/f-secure-foundry/tamago/soc/imx6/usb"
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/usbnet.go:24]: GH org

File: witness/golang/omniwitness/usbarmory/usbnet.go [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/golang/glog"
+	"github.com/miekg/dns"
+)
+
+const (
+	deviceIP  = "10.0.0.1"
+	deviceMAC = "1a:55:89:a2:69:41"
+	hostMAC   = "1a:55:89:a2:69:42"
+	resolver  = "8.8.8.8:53"
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/usbnet.go:33]: Might be worth popping in a TODO to think about DHCP?

File: witness/golang/omniwitness/usbarmory/usbnet.go [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize USB networking, %v", err)
+	}
+
+	iface.EnableICMP()
+
+	l, err := iface.ListenerTCP4(80)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize HTTP listener, %v", err)
+	}
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/usbnet.go:52]: Should we comment that out / ding it?

File: witness/golang/omniwitness/usbarmory/usbnet.go [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+				return nil, fmt.Errorf("failed to resolve A records for host %q", host)
+			}
+			var ip net.IP
+			if a, ok := r.Answer[0].(*dns.A); ok {
+				ip = a.A
+			} else {
+				return nil, fmt.Errorf("expected A record but got %T %q", r.Answer[0], r.Answer[0])
+			}
+			target := fmt.Sprintf("%s:%s", ip, port)
+			glog.Infof("Dialing %s", target)
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/usbnet.go:118]: nit: demote to `V(1)`?

File: witness/golang/omniwitness/usbarmory/usbnet.go [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+			return iface.DialTCP4(target)
+		},
+		TLSClientConfig: &tls.Config{
+			// TODO: determine some way to make client certs available
```
Comment (AlCutter) [inline witness/golang/omniwitness/usbarmory/usbnet.go:122]: I guess they could be embedded too right? Maybe the `Makefile` can figure out whether to look in `/etc/ssl/certificates` or somewhere for the trusted set?

**Review by AlCutter — COMMENTED** [review]
Few randos inline, nothing huge!

File: witness/golang/omniwitness/usbarmory/usbnet.go (hunk 19 above)
Comment (mhutchinson) [reply witness/golang/omniwitness/usbarmory/usbnet.go:122]: Yeah, there are ways and means of solving this. But this TODO is for the further quite a way beyond this PR :-)

**Review by mhutchinson — COMMENTED** [review]
Ta. I've left open the threads on which we didn't obviously agree. PTAL?

File: witness/golang/omniwitness/usbarmory/usbnet.go (hunk 16 above)
Comment (AlCutter) [reply witness/golang/omniwitness/usbarmory/usbnet.go:33]: I'm reasonably sure we will need to do this at some point, but happy to leave the TODO out - as you say, it'll be obvious when we need it.

File: witness/golang/omniwitness/usbarmory/main.go (hunk 12 above)
Comment (AlCutter) [reply witness/golang/omniwitness/usbarmory/main.go:49]: Might be worth popping a comment in for the spooky action at a distance?

File: witness/golang/omniwitness/usbarmory/usbnet.go (hunk 17 above)
Comment (AlCutter) [reply witness/golang/omniwitness/usbarmory/usbnet.go:52]: Ok, maybe I need to look at this code again - this appears to be opening a random listen on a hard-coded port 80, but the witness is listening on 8080 so I assumed this was unused cruft that came in with the usbnet code.

Comment (AlCutter) [reply witness/golang/omniwitness/usbarmory/usbnet.go:52]: Ah, that `8080` was the `monolith` main, my bad.

**Review by AlCutter — APPROVED** [review]

**Pull request merged**